### DISCUSSION TITLE
Proposal: Support bidirectional close tilt for covers

### DISCUSSION THREAD

**Discussion by raman325** [home-assistant/architecture#912] [Entity Models]
# Background

Both vertical binds and horizontal venetian blinds can tilt in two directions, which means that they have two states that could be considered closed. The tilt functionality in the `cover` platform does not handle this case and expects close tilt to be a single state.

# Proposal
- Add an additional feature for `CLOSE_TILT_ALTERNATE` which allows for a secondary tilt (blind slats can close in two directions)
- Consider adding an optional `direction` attribute with an enum that's either `UP_AND_DOWN` or `LEFT_AND_RIGHT`. This can be used by the frontend to decide how to display the `close_tilt` and `close_tilt_alternate` service
- add a `close_tilt_alternate` service or update the `close_tilt` service to accept an optional parameter that specifies direction somehow

In https://github.com/home-assistant/architecture/discussions/908 before this proposal was split out, @starkillerOG suggested alternate names for the new feature - I don't have a strong opinion on what the name should be. What I was thinking is that we preserve the existing features and services to handle closing in one direction, and add a new feature and service to handle the second direction, but I am open to alternate suggestions on that too.

**Comment by bdraco** [comment]
When we considering changing the entity model we need to know which integrations would benefit from the change since we want to avoid making architecture changes for one-off cases.

There are probably a lot of cases that could apply to https://www.home-assistant.io/integrations/hunterdouglas_powerview/ but I don't have any of these devices so I'm not the best person to give input here.

**Comment by ZephireNZ** [comment]
This issue has come up as part of implementing Close Up/Down in SwitchBot Blind Tilt: https://github.com/home-assistant/core/pull/96539

My initial approach was to simply allow user to configure the default close direction for the integration per device, but bdraco (rightly) pointed out that it should ideally be a UI feature and change to service calls.

Based on your own proposal, and discussions with bdraco, my proposal would be:

- Add a feature flag `CLOSE_TILT_MULTIDIRECTIONAL`
- Add an entity attribute `directions` that lists possible close directions. This mirrors the way eg `fan` works to support arbitrary preset modes.
- Add an optional parameter `direction` to service `cover.close_cover_tilt` so user can specify up/down/left/right/etc from above. Default behaviour when direction is not provided would be integration specific (as it currently is).
- In the UI, if the above feature flag is set then show buttons for each close mode (eg Close Up, Close Down, etc) based on the attribute options. This again mirrors the way that `fan` works in the UI for preset modes. We can mirror the way `alarm` works with the `tile` card and give user ability to show/hide buttons if needed.

**Comment by ZephireNZ** [comment]
> When we considering changing the entity model we need to know which integrations would benefit from the change

I did a review of integrations implementing `close_cover_tilt` and how its library handles the close command.

Integrations that would benefit from multidirectional close:
- [_**esphome**_](https://www.home-assistant.io/integrations/esphome) - currently [hardcodes](https://github.com/home-assistant/core/blob/62c5194bc8928eb8591ad07b61a28f4d0c05b0e1/homeassistant/components/esphome/cover.py#L119-L125) 0% as closed, 100% as open but ESPHome itself does not define what is open/closed
- [_**hunterdouglas_powerview**_](https://www.home-assistant.io/integrations/hunterdouglas_powerview) - library currently [hardcodes](https://github.com/sander76/aio-powerview-api/blob/8dceb352c864a24236782638544f0e9865af8421/aiopvapi/resources/shade.py#L128) tilt of 0% for closed and 50% for open
- [_**mqtt**_](https://www.home-assistant.io/integrations/mqtt) - MQTT currently defines only one closed position [based on config](https://github.com/home-assistant/core/blob/62c5194bc8928eb8591ad07b61a28f4d0c05b0e1/homeassistant/components/mqtt/cover.py#L615)
- [_**soma**_](https://www.home-assistant.io/integrations/soma) - SomaConnect [supports flag `close_upwards`](https://support.somasmarthome.com/hc/en-us/articles/360026064234-HTTP-API#:~:text=morning_mode%20%3D%200/1-,close_upwards,-specifies%20the%20closing) but this is not currently used in library
- [_**switchbot**_](https://www.home-assistant.io/integrations/switchbot) - library [supports close up/down](https://github.com/Danielhiversen/pySwitchbot/blob/2db1006e5f8be8212737b3c8eb4cff22eabc5a0b/switchbot/devices/blind_tilt.py#L55-L63) but it is unused
- [_**template**_](https://www.home-assistant.io/integrations/template) - would enable simpler custom venetian/vertical blinds with multidirectional close without custom scripts
- [_**velux**_](https://www.home-assistant.io/integrations/velux) - library currently [hardcodes](https://github.com/Julius2342/pyvlx/blob/2649f1c7ed221971a662bb272fbf653136571d23/pyvlx/opening_device.py#L177) tilt of 50% as open and 100% as close
- [_**zwave_js**_](https://www.home-assistant.io/integrations/zwave_js) - was recently changed to [use a close/open "target"](https://github.com/home-assistant/core/blob/62c5194bc8928eb8591ad07b61a28f4d0c05b0e1/homeassistant/components/zwave_js/cover.py#L239) from Zwave, could be expanded to support multiple close targets

Integrations that don't appear to support multidirectional close:
- [_**acmeda**_](https://www.home-assistant.io/integrations/acmeda) - not true tilt, uses same API as `close_cover`
- [_**bond**_](https://www.home-assistant.io/integrations/bond) - calls library API to close with no direction
- [_**deconz**_](https://www.home-assistant.io/integrations/deconz) - device does close tilt one-directional (0 open to 100 closed)
- [_**dynalite**_](https://www.home-assistant.io/integrations/dynalite) - calls library API to close with no direction
- [_**fibaro**_](https://www.home-assistant.io/integrations/fibaro) -  device does close tilt one-directional (100 open to 0 closed)
- [_**homematic**_](https://www.home-assistant.io/integrations/homematic) - device does close tilt one-directional (1.0 open to 0.0 closed)
- [_**homematicip_cloud**_](https://www.home-assistant.io/integrations/homematicip_cloud) -  device does close tilt one-directional (0.0 open to 1.0 closed)
- [_**knx**_](https://www.home-assistant.io/integrations/knx) - device does close tilt one-directional (0 open to 100 closed)
- [_**motion_blinds**_](https://www.home-assistant.io/integrations/motion_blinds) - library limits tilt rotation to 0-180 degrees, not sure if this is a device limitation or library 
- [_**overkiz**_](https://www.home-assistant.io/integrations/overkiz) - calls library API to close with no direction
- [_**rfxtrx**_](https://www.home-assistant.io/integrations/rfxtrx) - not true tilt, uses same API as `close_cover`
- [_**tasmota**_](https://www.home-assistant.io/integrations/tasmota) - calls library API to close with no direction

**Reply by starkillerOG** [reply]
Motionblinds tilt 0 is closed on one direction, 180 is closed the other direction and 90 degrees the slats are horizontal so open.

Bidirection would therefore benefit Motion Blinds.

**Reply by vexofp** [reply]
I can second that this would be useful for ZWave. I have a bunch of [iBlinds](https://myiblinds.com/) window blinds motors to which this would apply.

Specifically, these and similar devices expose the "Horizontal Slats Angle" property of ZWave's Window Covering command class. This property includes the notion of two "closed" states, one at either end with "open" being halfway in between ([`SlatsStates.CLOSED_1` and `.CLOSED_2` in `zwave-js-server-python`](https://github.com/home-assistant-libs/zwave-js-server-python/blob/02ebbd8f546838a68a7a689c77c33320d5cc496e/zwave_js_server/const/command_class/window_covering.py#L57)). The ZWave integration currently maps Home Assistant's notion of closed to [one of the two states, hard-coded arbitrarily](https://github.com/home-assistant/core/blob/62c5194bc8928eb8591ad07b61a28f4d0c05b0e1/homeassistant/components/zwave_js/cover.py#L398).

I think it would be great to expand Home Assistant's notion of closing to encompass such double-ended devices. The design proposal above seem like they would cover such devices well! 👍

**Reply by Lonestar10** [reply]
Has any of this been implemented in ZWave?  I cannot use [iBlinds](https://myiblinds.com/) properly because the state appears as "unknown."

**Reply by vexofp** [reply]
iBlinds state showing as unknown is actually due to #908 , and not in particular anything to do with being bidirectional. They only report "tilt" and not "position", and currently Home Assistant's cover entity state is based solely on position.

I tried to jump-start discussion of some different potential solutions over in that other thread, but never got much traction. 🤷 I also prototyped some of the different proposed solutions in some branches ([1C](https://github.com/home-assistant/core/compare/dev...vexofp:homeassistant-core:tilt-cover-1c), [2A](https://github.com/home-assistant/core/compare/dev...vexofp:homeassistant-core:tilt-cover-2a)). Personally, I have been running the 1C option at home for a few months now and generally happy with it.

**Comment by BradleyFord** [comment]
Smartwings blinds via Zigbee2MQTT are fixed in terms of 0 to 100.

But what is zero depends on if your blind rolls over the front or back of
the roller.

**Comment by lkostrowski** [comment]
I know it's a little old but I will add some context of mine

I have a facade venetial blinds, type C80, which means they can rotate 180deg (contrary to Z90 which are 90 degrees).

So, when blinds are going up, they are first closed in direction 1 and start moving up.
When blinds are going down, they first rotate 180 deg and move down.

No mateter the direction, I need to move them half full rotation to achieve 90 deg.

It's a little limiting - HA entites have 0-100 tilt, which forces me to either opt-out native blinds position, or to write custom logic around them

Actual use case:
- During the winter I want them to be closed direction that will lighten the room, still with a privacy
- During the summer I want the reverse, to add some shade

It would require changing the entity
- 0 closed
- 50 full open
- 100 closed

**Comment by wollew** [comment]
I would like to revive this proposal and therefore tried to write up everything how I think it would make sense to implement. Looking forward to feedback. 

# Updated Proposal: Support multiple tilt closing modes for covers

## Summary and motivation

Extend the Home Assistant `cover` entity to support covers where the tilt mechanism can be closed in more than one orientation.

This is common for venetian blinds and similar coverings. For example, a blind may have one closed position with the slats tilted upwards and another with the slats tilted downwards:

```text
0°              90°             180°
│────────────────│────────────────│
closed           open            closed
  ↑                                ↑
slats up                       slats down
```

Both 0° and 180° are closed, but they may have different practical effects:

- daylight penetration
- glare
- etc.


The current `cover` API can represent the resulting tilt positions, but it cannot express **which closed orientation the user wants**.

The proposal is to add the concept of **tilt closing modes**:

```python
tilt_closing_modes = ["up", "down"]
```

and extend `cover.close_cover_tilt` with an optional `closing_mode`:

```yaml
action: cover.close_cover_tilt
target:
  entity_id: cover.living_room_blind
data:
  closing_mode: down
```

The strings are stable machine-readable identifiers. Their human-readable representation is provided through Home Assistant's normal translation mechanism (`strings.json`).

The integration is responsible for mapping a closing mode to whatever the underlying device requires. For devices that do not natively report these capabilities, the proposal introduces a standardized way to enable multidirectional tilt support and configure the available modes directly on the entity via the **Entity Registry**.


---

## API 
### Entity changes

Add an optional entity property:

```python
class CoverEntity:
  @property
  def tilt_closing_modes(self) -> list[str] | None:
      """Return the supported tilt closing modes.
         If None or [], device doesn't support the closing mode parameter.
      """
      return None
```

An integration supporting two orientations could return:

```python
tilt_closing_modes = ["up", "down"]
```
The values are opaque identifiers to Home Assistant core. The integration defines their meaning and maps them to its device API.

### Service changes
Add an optional `closing_mode` string parameter to the `cover.close_cover_tilt` service.

### Feature discovery
A separate `CLOSE_TILT_MULTIDIRECTIONAL` feature flag is introduced to allow feature discovery in the frontend (see this discussion here why it's needed: https://github.com/home-assistant/core/pull/176231#discussion_r3805639580)

`CLOSE_TILT_MULTIDIRECTIONAL` should be set iff `tilt_closing_modes` contains multiple modes.

`tilt_closing_modes` must be included in the entity's state attributes so the frontend can read the available modes.

### Validation
- If `closing_mode` is not in `tilt_closing_modes`, raise `ServiceValidationError`.
- If parameter is not provided, use the device default (integration decides).
- Parameter needs to be removed from service calls for entities that do not support 

### Translations

The identifiers can be translated using Home Assistant's existing translation mechanism.

### Backwards compatibility

Existing service calls remain valid:

```yaml
action: cover.close_cover_tilt
target:
  entity_id: cover.living_room_blind
```

When no mode is supplied, the integration retains its existing/default behavior. This preserves backward compatibility. The proposal deliberately does not redefine what the default closing mode is.

---

`cover.open_cover_tilt` remains unchanged.

It continues to mean "Open the tilt of the cover."

The integration remains responsible for determining the appropriate open/neutral position.

This is already part of the integration's responsibility today: the integration has to know how its particular device implements `open_cover_tilt`.

For example, a device might have:

```text
0%    = closed
30%   = open
100%  = closed
```

while another might have:

```text
0%    = closed
100%  = open
```

`open_cover_tilt` does not need to know or expose these positions.

There is currently no known integration requiring multiple distinct *open* tilt states comparable to the multiple closed states described above. ([GitHub](https://github.com/home-assistant/architecture/discussions/912)), so this proposal is deliberately asymmetric:
- `open_cover_tilt` has one semantic operation: open the tilt.
- `close_cover_tilt` can optionally select between multiple closing modes.

---

## Configuration via Entity Registry

For integrations where the device API does not support discovering multidirectionality and/or the available closing modes, configuration shall be handled via the **Entity Registry**. 

To avoid the problem of feature discovery, a two-stage conditional configuration is proposed for the entity settings UI. This will require frontend changes but the alternative of handling this in an `OptionsFlow` would require every integration to reimplement something very similar, this proposal would handle it on a platform level.

### 1. Capability Trigger
The frontend will provide a boolean toggle: **"Supports multiple closing orientations"**.
*   **Storage:** Saved in the entity registry as `multidirectional_tilt`.
*   **Effect:** When enabled, the entity must report `CLOSE_TILT_MULTIDIRECTIONAL = True`, forcing the frontend to treat the device as multidirectional regardless of the hardware API.

### 2. Mode Selection
A multi-select list (e.g., "Up", "Down", "Inside", "Outside", etc.) will be displayed **only if** the "Supports multiple closing orientations" toggle is active.
*   **Storage:** Saved in the entity registry as `tilt_closing_modes`.
*   **Effect:** These values populate the `tilt_closing_modes` property of the entity.

### Implementation Logic
The integration shall resolve these properties by prioritizing the registry overrides:

```python
@property
def CLOSE_TILT_MULTIDIRECTIONAL(self) -> bool:
    # Triggered by hardware support OR user registry override
    return self._device_api.supports_multidirection() or \
           self.registry_entry.options.get("multidirectional_tilt", False)

@property
def tilt_closing_modes(self) -> list[str] | None:
    # 1. Priority: User-defined modes in the registry
    registry_modes = self.registry_entry.options.get("tilt_closing_modes")
    if registry_modes:
        return registry_modes
        
    # 2. Fallback: Hardware-reported modes
    return self._device_api.get_supported_modes()
```
The integration owns the mapping from those modes to the device API.

For example:

```text
"up"   → device position 0
"down" → device position 100
```

The numeric mapping therefore remains an integration implementation detail.

---

## Integrations benefiting

The original proposal identified the following integrations as benefiting from multidirectional tilt closing, most of them currently hardcoding some kind of support (single direction) instead of allowing full flexibility because their API does not allow for discovery: ([GitHub](https://github.com/home-assistant/architecture/discussions/912#discussioncomment-6455659))

- ESPHome
- Hunter Douglas PowerView
- Motion Blinds
- MQTT
- SOMA
- SwitchBot
- Template
- Velux
- Z-Wave JS
- Zigbee / SmartWings

Integrations that have only one possible closing mode require no changes.

## Non-goals
This proposal does not explicitly intend to solve other open discussions on the cover integration, like:
- support for tilt only devices (https://github.com/home-assistant/architecture/discussions/908), cover state will still depend on cover position
- additional use cases and device types: (https://github.com/home-assistant/architecture/discussions/979)

As far as I can see implementing this proposal does not stop any of those to be solved separately.

**Reply by AlCalzone** [reply]
> The proposal is to add the concept of **tilt closing modes**:
> ```
> tilt_closing_modes = ["up", "down"]
> ```

I think this definition is ambiguous. What defines "up"?
For this reason, the Z-Wave specs call the orientations "Closed, up inside" and "Closed, down inside". Granted, it is a bit of a mouthful, but doesn't leave the definition open to guessing.

While very niche, this proposal also forgets about vertical slats that close by turning the slats left/right.
Again borrowing from Z-Wave, the closed directions could be called "Closed to the left inside" and "Closed to the right inside".

And there is also a third orientation, where we might need even different names for the closed directions:
<img width="720" height="720" alt="image" src="https://github.com/user-attachments/assets/c83887a2-a5b3-422f-97d7-ef5aee3fbc34" />

**Reply by wollew** [reply]
Thanks for the feedback. The intention of the proposal is to let integrations define their own `tilt_closing_modes`, depending on the available covers. Also, I would see the ZWave definitions you mention to just be part of the translation strings, also integration defined.
